### ISSUE TITLE
Make libjit an easier cross-compilation target

### ISSUE THREAD

**Issue by petecoup** [pytorch/glow#2520]
Right now, libjit is built and embedded inside the CPU backend.  It is built using the native target, which matches the CPU backend.
One possibility to make this easier for cross-compilers:
Move libjit and the associated build rules somewhere under LLVMCodeGen.  Add additional cmake rules/options to specify additional non-native targets to build libjit for.
Hide the embedded bitcode behind an interface that can be looked up via target name.

The CPU backend should be able to use this new interface by requesting the "native" libjit bitcode.

Thoughts?

**Comment by opti-mix** [comment]
@petecoup I agree that it is interesting to have the bitcode of libjit available for different cross-compilation targets.

> One possibility to make this easier for cross-compilers:
Move libjit and the associated build rules somewhere under LLVMCodeGen. Add additional cmake rules/options to specify additional non-native targets to build libjit for.
Hide the embedded bitcode behind an interface that can be looked up via target name.
The CPU backend should be able to use this new interface by requesting the "native" libjit bitcode.

Hmm. I intentionally kept libjit inside the `CPUBackend` instead of `LLVMIRGen`. The reason for that is that not all backends which use LLVM want to use the same implementation of a `libjit` with the same set of provided functions. Some of the backends may want to provide their own kernels and do not need the common `libjit` kernels provided by the `CPUBackend`. But on the other hand, `LLVMIRGen` has a pretty intricate knowledge about the kernels in `libjit` because it needs to  generate code to invoke them... So, it seems like we may want either move the `libjit` from `CPUBackend` into `LLVMIRCodeGen` or move instruction generation parts from `LLVMIRGen` into `CPULLVMIRGen`. If a specific backend needs to use its own libjit with different kernels, it could just override `getLibjitBitcode` and instruction generation functions.

I don't know if we need to support a use-case, where a backend uses most of the common `libjit` kernels, but also has some kernels of its own. If this needs to be supported, we may end-up with multiple bitcode files (e.g. one for the common libjit and one for the backend-specific kernels). WDYT?

> Add additional cmake rules/options to specify additional non-native targets to build libjit for.

+1 on this. It should be possible to provide a list of architectures/targets to compile for (it could include for each target the target triple, cpu and cpu features).

If we have multiple bitcode files of libjit for different targets, we need to have an API to use the libjit bitcode matching the selected target currently used by the CPU/LLVM backend. `getLibjitBitcode` could be extended to take a target parameter and return a matching bitcode.

**Comment by petecoup** [comment]
@opti-mix: My most immediate goal here is to support a cross-compiler on an AOT compilation path.  This can likely be accomplished by leaving libjit where it is, doing cmake work, and a few mods like you describe (give `getLibjitBitcode` a target parameter).  The current structure in place supports that, and if that is a logical first step, I can do that.

> I don't know if we need to support a use-case, where a backend uses most of the common libjit kernels, but also has some kernels of its own. If this needs to be supported, we may end-up with multiple bitcode files (e.g. one for the common libjit and one for the backend-specific kernels). WDYT?

This is a very good question.
The current design using `LLVMBackend` would let us drop in any libjit implementation we want that satisfies the current interface in `CPUBackend/libjit'.  This is great.

Going further, it would be nice if a backend could:

- Have backend/target-specific versions of existing libjit kernels without needing an entire backend/target-specific libjit.

- "Automatically" get enabled with new libjit implementations enabled through `LLVMBackend`
 
Not sure if that means multiple bitcode files, or surgery to split libjit apart so it looks more like `libm`.

**Comment by opti-mix** [comment]
@petecoup 

> Have backend/target-specific versions of existing libjit kernels without needing an entire backend/target-specific libjit.

Do you have an idea how this should be done? If I understand you correctly, these specific versions of existing libjit kernels would override the "standard" implementations from libjit, right? How would such a mechanism be implemented? At bitcode level? Or would we have multiple bitcode files containing potentially multiple versions of the same kernel and there will be some kind of a mechanism to pick the "right" one? How?

> "Automatically" get enabled with new libjit implementations enabled through LLVMBackend

I'm not sure what you mean here.

> Not sure if that means multiple bitcode files, or surgery to split libjit apart so it looks more like libm.

You mean like having a separate bitcode file per kernel?

**Comment by petecoup** [comment]
> Do you have an idea how this should be done? If I understand you correctly, these specific versions of existing libjit kernels would override the "standard" implementations from libjit, right? How would such a mechanism be implemented? At bitcode level? Or would we have multiple bitcode files containing potentially multiple versions of the same kernel and there will be some kind of a mechanism to pick the "right" one? How?

> You mean like having a separate bitcode file per kernel?

This is practically what I mean by this.  In short, perhaps we'd have something like:
`libjit/libjit_max_pool_xy_grad_f.cpp`
`libjit/my-target/libjit_max_pool_xy_grad_f.cpp`
When building libjit.bc for my-target, we'll compile and link in `libjit/my-target/libjit_max_pool_xy_grad_f.cpp`.  When building libjit.bc for any other target, we'll compile and link in `libjit/libjit_max_pool_xy_grad_f.cpp`.
After this, libjit has tons of files and a much more elaborate build system.

What I had meant in the "Automatically enable" comment: if someone adds a libjit implementation for some new kernel, it would be nice if any Backend using `LLVMBackend` would get that implementation without having to copy it into their own libjit implementation, even if they had specialized some of their other routines.

Perhaps there are better ways to accomplish these, and I don't feel strongly that these are things needed now.  But, if these use-cases seem reasonable, I'm happy to help implement some solution.

In any case, my near-term goal is just to be able to cross-compile and use libjit as-is, and I'd be happy to accomplish just that.

**Comment by opti-mix** [comment]
> In any case, my near-term goal is just to be able to cross-compile and use libjit as-is, and I'd be happy to accomplish just that.

>This can likely be accomplished by leaving libjit where it is, doing cmake work, and a few mods like you describe (give getLibjitBitcode a target parameter). The current structure in place supports that, and if that is a logical first step, I can do that.

Yes, it seems like a reasonable short-term improvement. Please submit a PR and I'd be happy to review it.

**Comment by petecoup** [comment]
One thing here.
Maintaining the current structure here means we'll be embedding ~200K (by my measurements) per cross-target specified into binaries, through CPUBackend.o.
This isn't too crazy, but some people pay close attention to these sorts of things.  Is that going to be OK for the short term?

**Comment by opti-mix** [comment]
@petecoup 

> Maintaining the current structure here means we'll be embedding ~200K (by my measurements) per cross-target specified into binaries, through CPUBackend.o.
This isn't too crazy, but some people pay close attention to these sorts of things. Is that going to be OK for the short term?

We used to store those bitcode files as real files. But this had some issues with FB internal build system called `buck` and also with searching for those files at run time and thus at some point we decided to simply embed them into source files to avoid all those issues.  So far, there were no complaints about it. Should someone complain, I'm pretty sure we should be able to resurrect the file based solution and solve the issues I just mentioned.
